[VEN-2887] - Vaults(4626) implementation for vTokens

Debugger022 opened the pull request:
## Description

This PR introduces the implementation of the ERC-4626 wrapper for Venus markets for isolated pools, enabling seamless integration with external DeFi protocols that follow the ERC-4626 standard. This enhancement includes the following changes:

ERC-4626 Implementation: Developed an ERC-4626 compliant wrapper for Venus VToken markets, standardizing deposits, withdrawals, and balance calculations.

Reward Distribution: Integrated reward claiming from RewardsDistributor, transferring rewards to the designated recipient and handling updates for Protocol Share Reserve (PSR) contracts.

Factory Contract: Introduced a factory contract to facilitate the deployment of ERC-4626 wrappers for different Venus markets.

MockVenusERC4626 Contract: Added a mock implementation of the ERC-4626 wrapper for testing purposes, enabling controlled simulations of deposits, withdrawals, and reward interactions.

Test Cases: Added comprehensive unit tests covering deposits, withdrawals, reward claiming, and recipient scenarios (EOA vs. PSR).

Hardhat Config File : fixed the external deployment dependency of arbitrumone for governance-contracts

Debugger022 committed:
feat: [WIP] vaults(4626) implementation for vTokens

Debugger022 committed:
feat: erc 4626 factory contracts

Debugger022 committed:
test: add tests for the erc-4626 implementation

Debugger022 committed:
feat: add action pause functionality to 4626 vaults

chechu commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,260 @@
+    /// @param receiver The address of the account redeeming.
+    /// @return The maximum number of shares that can be redeemed.
+    function maxRedeem(address receiver) public view override returns (uint256) {
+        uint256 cash = VTOKEN.getCash();
```
Shouldn't we check if the REDEEM action is paused in the vToken (similar to the check we are performing in the deposit flow)?

chechu commented on contracts/ERC4626/Base/ERC4626Factory.sol:
```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.13;
```
0.8.25?

chechu commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,135 @@
+    /// -----------------------------------------------------------------------
+
+    /// @notice Maps underlying asset to the corresponding vToken
+    mapping(ERC20 => VToken) public underlyingToVToken;
```
This mapping is already stored in the `poolRegistry` associated with the comptroller. We could call something like `comptroller.poolRegistry().getVTokenForAsset(comptroller, asset)`, to get the vToken associated with the asset in that comptroller

This way, we could remove this mapping and the function `updateUnderlyingToVToken`

chechu commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,135 @@
+    /// -----------------------------------------------------------------------
+
+    /// @notice The XVS token contract
+    ERC20 public immutable XVS;
```
I think we don't need `XVS` and `REWARD_DISTRIBUTOR` as attributes in this factory

chechu commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,260 @@
+        if (xvsBalance > 0) {
+            XVS.safeTransfer(REWARD_RECIPIENT, xvsBalance);
+        }
+        emit ClaimRewards(xvsBalance);
```
I think we should send them to the ProtocolShareReserve, with a new Income type: `ERC4626_WRAPPER_REWARDS`. I wouldn't hardcode the PSR address, ofc, we can keep the `REWARD_RECIPIENT` attribute, but we could "try-catch" call `REWARD_RECIPIENT.updateAssetsState` at the end of this function

Debugger022 replied:
Okay, I am checking for the rewardRecipient contract size if it exist then I am calling the `REWARD_RECIPIENT.updateAssetsState`
[3306ff2](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/3306ff24862b3f0f8cb2a3544d6054571336d16e)

Debugger022 committed:
fix: pr comments

Debugger022 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,260 @@
+            revert VaultActionPaused(action);
+        }
+    }
+}
```
[3306ff2](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/3306ff24862b3f0f8cb2a3544d6054571336d16e)

Debugger022 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,260 @@
+
+        uint256 supplyCap = COMPTROLLER.supplyCaps(address(VTOKEN));
+        uint256 totalSupply_ = VTOKEN.totalSupply();
+        return supplyCap > totalSupply_ ? supplyCap - totalSupply_ : 0;
```
[3306ff2](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/3306ff24862b3f0f8cb2a3544d6054571336d16e)

Debugger022 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,260 @@
+
+        if (xvsBalance > 0) {
+            XVS.safeTransfer(REWARD_RECIPIENT, xvsBalance);
+        }
```
Added functionality to support multiple rewardDistributors, [3306ff2](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/3306ff24862b3f0f8cb2a3544d6054571336d16e)

Debugger022 committed:
Merge branch 'develop' into feat/VEN-2887

chechu commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -72,108 +56,106 @@ contract VenusERC4626 is ERC4626, Ownable {
     }
 
+    function totalAssets() public view virtual override returns (uint256) {
+        return (VTOKEN.balanceOf(address(this)) * VTOKEN.exchangeRateStored()) / 1e18;
```
What if we used `EXP_SCALE` from `lib/constants.sol` instead of hardcoding 1e18 in this file?

GitGuru7 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,244 @@
+                SafeERC20.safeTransfer(rewardToken, REWARD_RECIPIENT, rewardBalance);
+
+                // Try to update the asset state on the recipient if reward recipient is a protocol share reserve
+                if (REWARD_RECIPIENT.code.length != 0) {
+                    IProtocolShareReserve(REWARD_RECIPIENT).updateAssetsState(
+                        address(COMPTROLLER),
+                        address(rewardToken),
+                        IProtocolShareReserve.IncomeType.ERC4626_WRAPPER_REWARDS
+                    );
+                }
+            }
```
Would be better to do a low level static call to check if the function exists?

```
(bool success, ) = target.staticcall(abi.encodeWithSelector(selector));
return success;
```

This way the function does not have to revert if the reward recepient is a contract but not the PSR.

GitGuru7 commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,82 @@
+
+    /// @inheritdoc ERC4626Factory
+    function createERC4626(ERC20 asset) external virtual override returns (ERC4626 vault) {
+        VToken vToken = VToken(
+            PoolRegistryInterface(COMPTROLLER.poolRegistry()).getVTokenForAsset(address(COMPTROLLER), address(asset))
+        );
```
Maybe we can optimise the constructor args of venusErc4626 contract to accept address of vToken to avoid importing the whole vToken contract here.

GitGuru7 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,244 @@
+    /// @param vToken_ The Venus vToken contract.
+    /// @param rewardRecipient_ The address receiving the rewards.
+    constructor(
+        VToken vToken_,
```
maybe convert this to address to avoid importing the complete vToken contract in the factory.

GitGuru7 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -0,0 +1,244 @@
+    function claimRewards() external {
+        RewardsDistributor[] memory rewardDistributors = COMPTROLLER.getRewardDistributors();
+
+        for (uint256 i = 0; i < rewardDistributors.length; i++) {
```
Would make sense to analyse the gas cost here to ensure we never run out of gas for 3-4 reward distributors. This could become a problem if there are more added in the future.

GitGuru7 commented on package.json:
```diff
@@ -51,7 +51,8 @@
     "hardhat-deploy": "^0.11.14",
-    "module-alias": "^2.2.2"
+    "module-alias": "^2.2.2",
+    "solmate": "^6.8.0"
```
Are we using solmate?

Debugger022 replied:
Nope, removed, [2a4290e](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/2a4290ee5f4307d10cad51f0c0e2faf6fda31346)

Debugger022 commented on contracts/ERC4626/VenusERC4626.sol:
```diff
@@ -1,55 +1,39 @@
-import { RewardDistributorInterface } from "./Interfaces/IRewardDistributor.sol"; // Interface for claiming rewards.
-import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IComptroller, Action } from "./Interfaces/IComptroller.sol";
+import { RewardsDistributor } from ".././Rewards/RewardsDistributor.sol";
```
we are getting distributors through `COMPTROLLER.getRewardDistributors()` and it returns array of distributors. so we had to use `RewardsDistributor instead of `IRewardsDistributor`.

kkirka reviewed (commented):
1. My main concern wrt the proposed implementation is upgradeability. VenusERC4626 uses quite complex logic involving reward claims, supply/borrow caps, PSR integration, etc. Although we try to preserve interface compatibility between the protocol revisions, there is no guarantee we won't at some point introduce additional functionality that needs to be taken into account in this wrapper (e.g. per-user caps, defferent reward mechanics, more restrictions on mints and redemptions, etc.). A beacon proxy per ERC4626 could be a solution here (we could still use the create2 approach), but AFAIU we won't be able to use immutables in this case.

2. AFAIU, the proposal is to deploy a factory per pool. So it's neither a "one single factory for everything" nor a "separate contract per each vToken". I wonder why we opt for something in between here, as it seems we're taking the worst from both worlds – we  still need to deploy and maintain multiple contracts on chain _and_ we need to deal with create2 and its limitations. Maybe there's some flexibility  we want to achieve this way (like setting different reward receivers for different pools)?

chechu commented:
> 2. AFAIU, the proposal is to deploy a factory per pool. 

I think we could improve this with these changes:

* replace the `COMPTROLLER` attribute with a `POOL_REGISTRY` attribute (there is a pool registry per chain, so we would have a VenusERC4626 factory per chain)
* accept the VToken address in the `createERC4626` function of the factory. We would require that the VToken is valid in the pool registry

Debugger022 committed:
refactor: remove solmate package

Debugger022 committed:
refactor: use beacon proxy for ERC4626 and make factory upgradable

Debugger022 committed:
refactor: make 4626 beacon upgradable

Debugger022 committed:
test: refactor tests for 4626

Debugger022 committed:
feat: add script to deploy 4626 contract and factory

Debugger022 commented:
> 1. My main concern wrt the proposed implementation is upgradeability. VenusERC4626 uses quite complex logic involving reward claims, supply/borrow caps, PSR integration, etc. Although we try to preserve interface compatibility between the protocol revisions, there is no guarantee we won't at some point introduce additional functionality that needs to be taken into account in this wrapper (e.g. per-user caps, defferent reward mechanics, more restrictions on mints and redemptions, etc.). A beacon proxy per ERC4626 could be a solution here (we could still use the create2 approach), but AFAIU we won't be able to use immutables in this case.
> 2. AFAIU, the proposal is to deploy a factory per pool. So it's neither a "one single factory for everything" nor a "separate contract per each vToken". I wonder why we opt for something in between here, as it seems we're taking the worst from both worlds – we  still need to deploy and maintain multiple contracts on chain _and_ we need to deal with create2 and its limitations. Maybe there's some flexibility  we want to achieve this way (like setting different reward receivers for different pools)?

I have made ERC4626 contracts beacon upgradable and Factory contract is also upgradable now.
[091b01f](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/091b01f2cd6d9994c6a6519c140b7219af5e5c32)
[8e15e08](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/8e15e085f2d764a2e628fd99786a3edc01d8d3fa)

Debugger022 commented:
> > 2. AFAIU, the proposal is to deploy a factory per pool.
> 
> I think we could improve this with these changes:
> 
> * replace the `COMPTROLLER` attribute with a `POOL_REGISTRY` attribute (there is a pool registry per chain, so we would have a VenusERC4626 factory per chain)
> * accept the VToken address in the `createERC4626` function of the factory. We would require that the VToken is valid in the pool registry

1. Refactored.
2. I am using comptroller and asset address  in the `createERC4626` function of the factory.

Debugger022 committed:
fix: loopsLimit variable issue

kkirka commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,112 @@
+
+        vault = ERC4626Upgradeable(
+            address(
+                new BeaconProxy(
```
I thought the plan was to use a factory and create2, do you think using `new BeaconProxy{salt: salt}(...)` here can cause any issues?

Debugger022 replied:
Initially, I was thinking to use the factory and create2 for the 4626 vaults but to make vaults upgradable, I chose to opt for the beacon approach in the same way we are doing it with vTokens and comptrollers.
I guess it would not cause any issue, please let me know if you have any thoughts.

chechu commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,112 @@
+    /// @param comptroller The comptroller's address to which asset belongs
+    /// @param asset The ERC20 asset address belongs to the comptroller
+    /// @return vault The deployed VenusERC4626 vault
+    function createERC4626(address comptroller, address asset) external returns (ERC4626Upgradeable vault) {
```
Wouldn't it be more intuitive if we receive the address of a VToken here. This factory will deploy VenusERC4626 contracts, and each VenusERC4626 contract is associated with a VToken

chechu commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,112 @@
+/// @title VenusERC4626Factory
+/// @notice Factory for creating VenusERC4626 contracts
+contract VenusERC4626Factory is Ownable2StepUpgradeable, AccessControlledV8 {
+    using SafeERC20Upgradeable for ERC20Upgradeable;
```
Do we need this? I don't see any use of `ERC20Upgradeable` in this contract

Debugger022 replied:
Removed, [df78de4](https://github.com/VenusProtocol/isolated-pools/pull/497/commits/df78de4e335ede29fe3a680af369626be05aa1e7)

Debugger022 commented on contracts/ERC4626/Interfaces/IProtocolShareReserve.sol:
```diff
@@ -0,0 +1,13 @@
+    enum IncomeType {
+        SPREAD,
+        LIQUIDATION,
+        ERC4626_WRAPPER_REWARDS
```
Yes, I will raise an PR for the same in the reserve-protocol repo.

Debugger022 committed:
fix: pr comments

chechu commented on contracts/ERC4626/VenusERC4626Factory.sol:
```diff
@@ -0,0 +1,104 @@
+
+        VTokenInterface vToken_ = VTokenInterface(vToken);
+
+        if (poolRegistry.getPoolByComptroller(address(vToken_.comptroller())).comptroller == address(0)) {
```
I would check this:

```
vToken == poolRegistry.getVTokenForAsset(vToken_.comptroller(), vToken_.underlying());
```

To be sure the provided vToken is included in a valid pool registered in our registry. With the current check, any contract returning a valid comptroller in the `comptroller()` function would pass the check.

Debugger022 replied:
Used `getVTokenForAsset` instead of `getPoolByComptroller` to verify.

chechu commented on contracts/ERC4626/Interfaces/IComptroller.sol:
We could simply define the functions we'll use:

```
function actionPaused(address market, Action action) external view returns (bool);
function getRewardDistributors() external view returns (RewardsDistributor[] memory);
function supplyCaps(address) external view returns (uint256);
```

I wouldn't redefine the Action enum. We could import and use it from `contracts/ComptrollerInterface.sol`. Less code should be better :) wdyt?

Debugger022 committed:
refactor: deploy erc4626 vaults with create2

Debugger022 committed:
refactor: add constructor to 4626 vaults and factory

Debugger022 committed:
test: add tests for mint and redeem operations

Debugger022 committed:
refactor: add ReentrancyGuard to 4626 vaults

Debugger022 committed:
fix: initialize ReentrancyGuard

Debugger022 committed:
Merge pull request #504 from VenusProtocol/feat/VEN-3177

chechu committed:
Merge branch 'develop' into feat/VEN-2887

chechu committed:
docs: vew-12

chechu committed:
feat: vew-10

chechu committed:
feat: vew-07

* add ACM to the VenusERC4626 contract, and setter functions for the
maxLoopsLimit and rewardRecipient attributes

* add a second initializer to the VenusERC4626 contract, so the first
initializer can be invoked only providing the VToken address, and therefore
the CREATE2 from the Factory will calculate the same address for two vaults
wrapping the same VToken, not considering the rest of the parameters

* store a mapping in the Factory with the created VenusERC4626 contracts
per VToken

chechu committed:
feat: vew-06

Remove the total reserves of the VToken from the cash to calculate the
available cash to be considered in the maxmaxWithdraw and maxRedeem functions

chechu committed:
feat: vew-05

add sweepToken function to recover excess amount of underlying token redeemed
or any other token transferred to the VenusERC4626 tokens by mistake

chechu committed:
feat: vew-03

Set the owner of the deployed beacon contract to the owner of the Factory when
it's initialized

chechu committed:
Merge pull request #516 from VenusProtocol/feat/VEN-3191-jesus

[VEN-3191]: Mitigations considering the Certik audit report

Debugger022 committed:
fix: vew-04 and vew-09

chechu committed:
feat: m03

Debugger022 committed:
feat: L02 and L03

Debugger022 committed:
refactor: remove redundant checks in intialize2

Debugger022 committed:
feat: M02

chechu committed:
feat: l04

chechu committed:
Merge pull request #518 from VenusProtocol/feat/VEN-2887-jesus

[VEN-2887]: Mitigations on ERC4626 wrapper

chechu committed:
feat: adjust the assets to be multiples of the VToken exchange rates

chechu committed:
docs: add audit report for the ERC 4626 wrappers

Debugger022 committed:
chore: fix lint

Debugger022 committed:
test: refactor unit tests

Debugger022 committed:
test: Fork test for Venus4626

Debugger022 committed:
feat: vew-13

chechu committed:
test: add more fork tests for the ERC4626 wrapper

chechu committed:
Merge pull request #520 from VenusProtocol/feat/VEN-2887-round

[VEN-2887]: adjust the assets to be multiples of the VToken exchange rates

chechu committed:
feat: round up the deposited funds in the mint flow

Debugger022 committed:
feat: vew-11

Debugger022 committed:
fix: lint

chechu committed:
Merge branch 'develop' into feat/VEN-2887

chechu committed:
feat: vew-17

chechu committed:
refactor: vew-16

chechu committed:
feat: vew-17

Debugger022 committed:
feat: add sepolia deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

chechu committed:
feat: vew-04 - return actual shares after depositing

chechu committed:
docs: vew-10

chechu committed:
feat: vew-04 - consider imvolved vTokens in withdraw and redeem flows

chechu committed:
refactor: format internal function and move some checks

chechu committed:
feat: revert if VToken balance doesn't change during deposits and withdrawals

Debugger022 committed:
feat: add testnet deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
fix: redeploy sepolia ERC4626Factory

Debugger022 committed:
feat: add zksyncsepolia deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
fix: use transferOwnership in ERC4626Factory deployment script

Debugger022 committed:
feat: add arbitrumone and opmainnet deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
feat: add ethereum and basemainnet deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
feat: add unichainmainnet and zksyncmainnet deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
feat: add opbnbmainnet deployments for ERC4626Factory

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
fix: update the external deployment dependency for arbitrumone in hardhat.config.ts

Debugger022 committed:
Merge pull request #521 from VenusProtocol/feat/VEN-2887-certik

[VEN-2887]: mitigations for VenusERC4626 considering Certik audit

Debugger022 committed:
fix: update VenusERC4626Implementation deployments for testnets

Debugger022 committed:
Merge branch 'develop' into feat/VEN-2887-deployments

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
fix: update VenusERC4626Implementation deployments for mainnet

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
test: fix tests for 4626 vault

chechu committed:
docs: add certik audit of erc 4626 wrappers

chechu committed:
Merge branch 'develop' into feat/VEN-2887

chechu committed:
Merge branch 'feat/VEN-2887' into feat/VEN-2887-deployments

Debugger022 committed:
feat: add deployments for ERC4626Factory on bscmainnet and bsctestnet

Debugger022 committed:
feat: updating deployment files

chechu committed:
Merge pull request #523 from VenusProtocol/feat/VEN-2887-deployments

[VEN-2887]: add deployments for ERC4626Factory

chechu commented:
This code was moved to its own repository: https://github.com/VenusProtocol/ERC4626
The initial commit of that repository contains the codebase of this pull request: https://github.com/VenusProtocol/ERC4626/commit/771f392e011841a60fee68abe31cbf225b628b91

Debugger022 commented:
closing it: https://github.com/VenusProtocol/isolated-pools/pull/497#issuecomment-3056665824
